rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -1117,68 +1118,86 @@ extern "C" rmw_ret_t rmw_init(const rmw_init_options_t * options, rmw_context_t
 {
   rmw_ret_t ret;
 
-  static_cast<void>(options);
-  static_cast<void>(context);
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(options, RMW_RET_INVALID_ARGUMENT);
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(context, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    options->implementation_identifier,
+    "expected initialized init options",
+    return RMW_RET_INVALID_ARGUMENT);
   RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
     options,
     options->implementation_identifier,
     eclipse_cyclonedds_identifier,
     return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  if (NULL != context->implementation_identifier) {
+    RMW_SET_ERROR_MSG("expected a zero-initialized context");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
 
   if (options->domain_id >= UINT32_MAX && options->domain_id != RMW_DEFAULT_DOMAIN_ID) {
     RCUTILS_LOG_ERROR_NAMED(
       "rmw_cyclonedds_cpp", "rmw_create_node: domain id out of range");
     return RMW_RET_INVALID_ARGUMENT;
   }
 
+  rmw::impl::cpp::atexit cleanup{[context]() {
```

Comment:
This cleanup routine is written on the assumption that nothing that may require cleaning up will happen following the allocation of `context->impl`, which is true now. I am a little concerned that this model increases the risk of failure to cleanup correctly when additional resource claims are added in the initialization code at a later stage.

I don't perceive that "increase of risk" (if indeed there is any) to be large. To me, the exact way in which this arranges the cleanup doesn't really add anything of value, but it's not like I am against it. Clearly, you have a different sense of the value or you wouldn't have put in the effort. On balance, that makes me in favour of merging it, and I am raising this merely because it could be a perspective you hadn't taken into account yet.

Reply:
> To me, the exact way in which this arranges the cleanup doesn't really add anything of value, but it's not like I am against it. 

The sole point of using RAII for cleanup here is to de-duplicate cleanup code, with a preference for stack unwinding in lieu of `goto` statements.

Reply:
I like the RAII bit very much (one of the things that is so much better in C++ than in C ...), but to me it would then be more logical to do set it up immediately after the allocation happened. Anyway, as I said, I'm ok with it, especially after your reply.

Reply:
> but to me it would then be more logical to do set it up immediately after the allocation happened.

[Per API contract](https://github.com/ros2/rmw/pull/243/files#diff-9fe3987dc902dde2f416ab8fa78d0ed4R66) I must ensure that the context remains zero initialized if initialization fails not due to invalid use. So I have to have that block there to zero re-initialize it, and figured a no-op `delete` didn't make any harm. But yeah, we could split those two statements into separate `rcpputils::scope_exit` instances.